### PR TITLE
Tensorrt Mix Precision or INT8 conversion, mix precision almost same size and speed with INT8, but better precision, the converted model have good detection result with mix precision.

### PR DESCRIPTION
**Pull request by ZouJiu1**
1. I have checked for Existing Contributions: Before submitting, my contribution is unique and complementary.

2. No related issues.

new feature 

running model in edge device like nvidia jetson

3. **Purpose**:  I want to running yolov8x.pt in nvidia jetson orix nx 16G device，So I have to convert the trained yolov8x.pt model to tensorrt model. FP16 conversion and inference both are OK, it speed is faster than before, but I want more higher fps. 

So I want to convert it to int8, but when I add codes to ultralytics and convert yolov8x.pt to a int8 engine, it is still ok, but when inference with int8, I find it has no detection result. 

I search so many blogs and github issues, including tensorrt github, I was looking for a reason and a solution, but I can not find one. I try all the official yolov8*.pt with int8, these converted int8 engine have no detection result. But I find a possible solution, It is the mix precision [https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet/build_engine.py#L188-L218](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet/build_engine.py#L188-L218). Mix precision definition is that some layers are FP16, other layers are INT8.

When I use the mix precision to convert all official yolov8*.pt, **keep the first, second, last convolution layers FP16, other layers INT8**, these converted engines have a good detection result in bus.jpg, and it size just bigger a little 200KB, because just three convolution layer use FP16 and others still use int8. So my codes is right, my conversion and calibration procedure are correct. I think yolov8 can add the mix precision to convert model for a better performance.

i. The procedure of converting **a mix precision** engine：

1. download coco [val_2017.zip](http://images.cocodataset.org/zips/val2017.zip) and unzip to val_2017.

2. set calib_input=./val_2017, cache_file=./calibration.cache, **half=True, int8=True**

3. run the script, it will using the mix precision mode convert the model to engine file and infer with bus.jpg.<br>
before running, I marked the 322 line `raise SyntaxError(string + CLI_HELP_MSG) from e` in file ultralytics\\cfg\\__init__.py.

ii. The procedure of converting a **INT8** engine：all same beside this (half=False, int8=True)

the converted models (INT8 or MIX Precision) can download from here [https://www.alipan.com/s/FdfFoPDGCWH](https://www.alipan.com/s/FdfFoPDGCWH) in tensorrt 10.0.0b6. \*mix is a mix precision engine, \*int8 is a INT8 engine, nod is no detection.

mix precision log
```
TensorRT: input "images" with shape(1, 3, 640, 640) DataType.HALF
TensorRT: output "output0" with shape(1, 84, 8400) DataType.HALF
Mixed-Precision Layer /model.0/conv/Conv set to HALF STRICT data type
Mixed-Precision Layer /model.1/conv/Conv set to HALF STRICT data type
Mixed-Precision Layer /model.22/dfl/conv/Conv set to HALF STRICT data type
TensorRT: building FP16 engine as yolov8x.engine
TensorRT: building INT8 engine as yolov8x.engine
```

------------------------------------------------------------------------------------------------------------------
these engines using mix precision have a good detection result in bus.jpg, and these engines using INT8 have no result.

### related pull request [https://github.com/ultralytics/ultralytics/pull/9840](https://github.com/ultralytics/ultralytics/pull/9840) 

int8 conversion reference
[https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/detectron2)

mix precision conversion reference (much better)<br>
Tips: the firstly layer and last layer should be FP16, others can be int8. just try it.
[https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet](https://github.com/NVIDIA/TensorRT/blob/main/samples/python/efficientdet)

``` python []
import os
import gc
import sys
sys.path.append(r'E:\work\codeRepo\deploy\common\zj\ultralytics')
from ultralytics import YOLO # newest version from "git clone and git pull"
os.environ['CUDA_VISIBLE_DEVICES'] = '0'

'''
Platform: Window11

Ultralytics YOLOv8.1.44 🚀 Python-3.9.18 torch-2.2.1+cu118 CUDA:0 (NVIDIA GeForce RTX 4070 Ti, 12282MiB)

onnx 1.16.0 opset 17

TensorRT 10.0.0b6:  
https://developer.nvidia.com/downloads/compute/machine-learning/tensorrt/10.0.0/zip/TensorRT-10.0.0.6.Windows10.win10.cuda-11.8.zip

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:41:10_Pacific_Daylight_Time_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0
'''

if __name__ == '__main__':
    file = r'yolov8x.pt'
    model = YOLO(file)  # load a pretrained model (recommended for training)
    calib_input = r'E:\work\codeRepo\deploy\jz\val2017'
    cache_file = r'E:\work\codeRepo\deploy\jz\calibration.cache'
    if os.path.exists(cache_file):
        os.remove(cache_file)
    results0 = model.export(format='engine', simplify=True,
                            half=True,
                            int8=True,
                            calib_batch_size=20,
                            calib_num_images=len(os.listdir(calib_input)),
                            calib_input=calib_input,
                            cache_file=cache_file,
                            device='cuda:0')
    del model
    gc.collect()
    model = YOLO(r"E:\work\%s"%(file.replace(".pt", ".engine")))
    result = model.predict(
                           'https://ultralytics.com/images/bus.jpg', 
                           save_dir=r'.//', 
                           save=True)
```

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced ultralytics TensorRT exporting with INT8 and mixed precision support for better performance and efficiency on NVIDIA devices.

### 📊 Key Changes
- Added support for exporting models to TensorRT engine files with INT8 precision, providing higher inference speed while maintaining accuracy.
- Implemented mixed precision (combining INT8 and FP16) for certain layers to further optimize performance without significant loss in accuracy.
- Adjusted memory allocations for TensorRT engines to accommodate changes in workspace size requirements, particularly for TensorRT version 10 and above.
- Introduced calibration for INT8 conversion to ensure accuracy, including a dynamic method to manage calibration batches and image preprocessing.
- Provided detailed documentation and examples for converting models to TensorRT with INT8 and mixed precision modes, aimed at users targeting NVIDIA Jetson devices or similar.

### 🎯 Purpose & Impact
- **Faster Inference Speeds:** Users can now convert and run their models with INT8 precision, significantly increasing inference speeds on supported NVIDIA GPUs and edge devices.
- **Efficient Resource Use:** Mixed precision allows for fine-tuned control over model precision, achieving an optimal balance between speed, accuracy, and memory usage.
- **Broader Compatibility:** With added support for the latest TensorRT version 10, developers have more flexibility and can leverage the newest NVIDIA optimizations.
- **Accessible Performance Gains:** The changes democratize access to advanced optimizations for the community, enabling users with varying levels of expertise to enhance their models' performance on NVIDIA hardware.

Expect these enhancements to open up new applications and opportunities, particularly in edge computing scenarios where efficiency and speed are paramount. 🚀

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR appears to be a placeholder or mistake - no changes detected! 🤷

### 📊 Key Changes
- None. The PR diff is empty, indicating no code has been modified, added, or removed.

### 🎯 Purpose & Impact
- Since there are no changes, there's no direct impact or new benefits to users from this PR.
- It could serve as a reminder to double-check before submitting PRs to ensure meaningful contributions. 📝✅

In short, this is a no-op PR, likely submitted in error or for purposes not directly related to software development. Always a good time for a reminder about the importance of checking our work! 👀👍

### CODE REVIEW

**Comment by ZouJiu1** [comment]
> @ZouJiu1 thanks for the PR!
> 
> We've just introduced TRT 10 compatibility in #9516 which includes some of your earlier updates in a separate PR.
> 
> Can you please update this PR to main as a first step?

I have updated this pull request to the main branch, you can check it out in [https://github.com/ultralytics/ultralytics/pull/9969](https://github.com/ultralytics/ultralytics/pull/9969).

**Comment by glenn-jocher** [comment]
Thanks for the update, @ZouJiu1! 🚀 We'll review the changes on your updated PR shortly. Keep up the great work!
